discovery/scaleway/baremetal.go

```diff
@@ -0,0 +1,176 @@
+	baremetalIPVersionLabel  = baremetalLabelPrefix + "ip_version"
+	baremetalNameLabel       = baremetalLabelPrefix + "name"
+	baremetalOSNameLabel     = baremetalLabelPrefix + "os_name"
+	baremetalOSVersionLabel  = baremetalLabelPrefix + "os_version"
+	baremetalStatusLabel     = baremetalLabelPrefix + "status"
+	baremetalTagsLabel       = baremetalLabelPrefix + "tags"
+	baremetalTypeLabel       = baremetalLabelPrefix + "type"
+	baremetalZoneLabel       = baremetalLabelPrefix + "zone"
+)
+
+func newBaremetalDiscovery(conf *SDConfig) (*baremetalDiscovery, error) {
+	d := &baremetalDiscovery{
+		Port:      conf.Port,
+		Zone:      conf.Zone,
+		Project:   conf.Project,
+		AccessKey: conf.AccessKey,
+		SecretKey: string(conf.SecretKey),
+		Name:      conf.FilterName,
+		Tags:      conf.FilterTags,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "scaleway_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	profile, err := LoadProfile(conf)
+	if err != nil {
+		return nil, err
+	}
+	d.Client, err = scw.NewClient(
+		scw.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		scw.WithUserAgent(fmt.Sprintf("Prometheus/%s", version.Version)),
+		scw.WithProfile(profile),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up scaleway client: %w", err)
+	}
+
+	return d, nil
+}
+
+func (d *baremetalDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	api := baremetal.NewAPI(d.Client)
+
+	req := &baremetal.ListServersRequest{
+		Zone: scw.Zone(d.Zone),
+	}
+
+	if d.Project != "" {
+		req.ProjectID = scw.StringPtr(d.Project)
+	}
+
+	if d.Name != "" {
+		req.Name = scw.StringPtr(d.Name)
+	}
+
+	if d.Tags != nil {
+		req.Tags = d.Tags
+	}
+
+	servers, err := api.ListServers(req, scw.WithAllPages(), scw.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	offers, err := api.ListOffers(&baremetal.ListOffersRequest{
+		Zone: req.Zone,
+	}, scw.WithAllPages())
+	if err != nil {
+		return nil, err
+	}
+
+	osFullList, err := api.ListOS(&baremetal.ListOSRequest{
+		Zone: req.Zone,
+	}, scw.WithAllPages())
+	if err != nil {
+		return nil, err
+	}
+
+	var targets []model.LabelSet
+	for _, server := range servers.Servers {
+		labels := model.LabelSet{
+			baremetalIDLabel:     model.LabelValue(server.ID),
+			baremetalNameLabel:   model.LabelValue(server.Name),
+			baremetalZoneLabel:   model.LabelValue(server.Zone.String()),
+			baremetalStatusLabel: model.LabelValue(server.Status),
+		}
+
+		for _, offer := range offers.Offers {
+			if server.OfferID == offer.ID {
+				labels[baremetalTypeLabel] = model.LabelValue(offer.Name)
+			}
+		}
+
+		if server.Install != nil {
+			for _, os := range osFullList.Os {
+				if server.Install.OsID == os.ID {
+					labels[baremetalOSNameLabel] = model.LabelValue(os.Name)
+					labels[baremetalOSVersionLabel] = model.LabelValue(os.Version)
+				}
```

Comment:
you can `break` here, but is it really needed to have the full os id? what is os.ID? if that it a slug, that would be acceptable.

Reply:
Yes because there is no slug of the operating system in the server type. You need to do an additional query to an os list to match the os id that is present in the serverinstall